drivers/xen/emg_console/emg_console.c

```diff
@@ -71,18 +71,42 @@
  * DEALINGS IN THE SOFTWARE.
  */
 
-#include <uk/essentials.h>
+#include <uk/console/driver.h>
 #include <common/hypervisor.h>
+#include <uk/boot/earlytab.h>
+#include <uk/prio.h>
+#include <uk/plat/common/bootinfo.h>
 
-#define __XEN_CONSOLE_IMPL__
-#include "emg_console.h"
-
-int emg_console_output(const char *str, unsigned int len)
+__ssz console_out(struct uk_console *dev __unused, const char *str, __sz len)
 {
 	int rc;
 
 	rc = HYPERVISOR_console_io(CONSOLEIO_write, len, DECONST(char *, str));
 	if (unlikely(rc < 0))
 		return rc;
+
 	return len;
 }
+
+__ssz console_in(struct uk_console *dev __unused, char *str __unused,
```

Comment:
`ukconsole` is written so that a `__NULL` op is safe as long as the flag to use that operation isn't set. But I think drivers should generally provide both functions.

Reply:
In fact, I'd rather forbid setting an op to `__NULL` altogether.